Source/WebCore/Modules/url-pattern/URLPattern.cpp

```diff
@@ -26,29 +26,247 @@
 #include "config.h"
 #include "URLPattern.h"
 
+#include "URLPatternCanonical.h"
 #include "URLPatternInit.h"
 #include "URLPatternOptions.h"
 #include "URLPatternResult.h"
 #include <wtf/RefCounted.h>
+#include <wtf/URL.h>
 
 namespace WebCore {
 
 WTF_MAKE_TZONE_OR_ISO_ALLOCATED_IMPL(URLPattern);
 
-Ref<URLPattern> URLPattern::create(URLPatternInput&&, String&& baseURL, URLPatternOptions&&)
+template<typename CharacterType>
+static String escapeStringForCharacters(std::span<const CharacterType> characters)
+{
+    static constexpr std::array<LChar, 9> escapeCharacters = { '+', '*', '?', ':', '(', ')', '\\', '{', '}' }; // NOLINT
+
+    StringBuilder result;
+    result.reserveCapacity(characters.size());
+
+    for (auto& character : characters) {
+        if (std::find(escapeCharacters.begin(), escapeCharacters.end(), character) != escapeCharacters.end())
+            result.append('\\');
+
+        result.append(character);
+    }
+
+    return result.toString();
+}
+
+// https://urlpattern.spec.whatwg.org/#escape-a-pattern-string
+static String escapeString(StringView input)
+{
+    ASSERT(input.containsOnlyASCII());
+
+    if (input.is8Bit())
+        return escapeStringForCharacters(input.span8());
+
+    return escapeStringForCharacters(input.span16());
+}
+
+// https://urlpattern.spec.whatwg.org/#process-a-base-url-string
+static String processBaseURLString(StringView input, BaseURLStringType type)
+{
+    if (type != BaseURLStringType::Pattern)
+        return input.toString();
+
+    return escapeString(input);
+}
+
+// https://urlpattern.spec.whatwg.org/#process-a-urlpatterninit
+static ExceptionOr<URLPatternInit> processInit(URLPatternInit&& init, BaseURLStringType type, String&& protocol, String&& username,
+    String&& password, String&& hostname, String&& port, String&& pathname, String&& search, String&& hash)
+{
+    URLPatternInit result;
+    result.protocol = protocol;
```

Comment:
Can't we `WTFMove()` here and for all the non-init parameters below? Are they used later in the function?